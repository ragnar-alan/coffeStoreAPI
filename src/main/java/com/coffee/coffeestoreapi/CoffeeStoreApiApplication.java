package com.coffee.coffeestoreapi;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class CoffeeStoreApiApplication {

    public static void main(String[] args) {
        SpringApplication.run(CoffeeStoreApiApplication.class, args);
    }

}
