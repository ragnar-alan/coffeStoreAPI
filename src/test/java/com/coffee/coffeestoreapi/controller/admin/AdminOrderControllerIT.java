package com.coffee.coffeestoreapi.controller.admin;

import com.coffee.coffeestoreapi.BaseIT;
import io.restassured.RestAssured;
import org.approvaltests.core.Options;
import org.approvaltests.scrubbers.Scrubbers;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.http.MediaType;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.utility.DockerImageName;

import java.io.IOException;

import static org.approvaltests.JsonApprovals.verifyJson;

public class AdminOrderControllerIT extends BaseIT {
    private static final DockerImageName IMAGE_NAME = DockerImageName
            .parse("postgres:16-alpine")
            .asCompatibleSubstituteFor("postgres");
    @Container
    static final PostgreSQLContainer<?> POSTGRES_CONTAINER = new PostgreSQLContainer<>(IMAGE_NAME)
            .withDatabaseName("testdb")
            .withUsername("testuser")
            .withPassword("testpass");

    @LocalServerPort
    private int port;

    static {
        POSTGRES_CONTAINER.start();
    }

    @DynamicPropertySource
    static void overrideProperties(DynamicPropertyRegistry registry) {
        registry.add("spring.datasource.url", POSTGRES_CONTAINER::getJdbcUrl);
        registry.add("spring.datasource.username", POSTGRES_CONTAINER::getUsername);
        registry.add("spring.datasource.password", POSTGRES_CONTAINER::getPassword);
    }

    @BeforeEach
    void setUp() {
        RestAssured.port = port;
    }

    @Test
    void testGetOrderShouldPass() {
        var result = RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/orders/{orderNumber}", "ORD-1001")
                .then()
                    .statusCode(200)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testGetOrderShouldFail_dueToMissingResource() {
        RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/orders/{orderNumber}", "prkr")
                .then()
                    .statusCode(404)
                    .extract()
                    .asString();
    }

    @Test
    void testGetOrdersShouldPass() {
        RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/orders/list")
                .then()
                    .statusCode(200)
                    .extract()
                    .asString();
    }

    @Test
    void testPatchOrdersShouldPass() throws IOException {
        var result = RestAssured
                .given()
                    .contentType(MediaType.APPLICATION_JSON_VALUE)
                    .body(getFileContents("mockRequests/mockPatchOrderRequest.json"))
                .when()
                    .patch("/api/v1/admin/orders/{orderNumber}", "ORD-1004")
                .then()
                    .statusCode(200)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testPatchOrdersShouldFail_dueToRequestObjectValidationMissingOrderLines() throws IOException {
        var result = RestAssured
                .given()
                    .contentType(MediaType.APPLICATION_JSON_VALUE)
                    .body(getFileContents("mockRequests/mockPatchOrderRequestMissingOrderLines.json"))
                .when()
                    .patch("/api/v1/admin/orders/{orderNumber}", "ORD-1004")
                .then()
                    .statusCode(400)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testPatchOrdersShouldFail_dueToTooLongOrdererName() throws IOException {
        var result = RestAssured
                .given()
                    .contentType(MediaType.APPLICATION_JSON_VALUE)
                    .body(getFileContents("mockRequests/mockPatchOrderRequestTooLongOrdererName.json"))
                .when()
                    .patch("/api/v1/admin/orders/{orderNumber}", "ORD-1004")
                .then()
                    .statusCode(400)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testPatchOrdersShouldFail_dueToNotExistingOrder() throws IOException {
        var result = RestAssured
                .given()
                    .contentType(MediaType.APPLICATION_JSON_VALUE)
                    .body(getFileContents("mockRequests/mockPatchOrderRequest.json"))
                .when()
                    .patch("/api/v1/admin/orders/{orderNumber}", "misssing-order-number")
                .then()
                    .statusCode(404)
                    .extract()
                    .asString();
        verifyJson(result);
    }

    @Test
    void testDeleteOrdersShouldPass() {
        RestAssured
                .given()
                .when()
                    .delete("/api/v1/admin/orders/{orderNumber}", "ORD-1010")
                .then()
                    .statusCode(204)
                    .extract()
                    .asString();

        //Verifying that the order is soft-deleted (status is set to cancelled and the cancelled_at date is populated)
        var result = RestAssured
                .given()
                .when()
                    .get("/api/v1/admin/orders/{orderNumber}", "ORD-1010")
                .then()
                    .statusCode(200)
                    .extract()
                    .asString();
        verifyJson(result, new Options().withScrubber(DATE_TIME_SCRUBBER));
    }
}
