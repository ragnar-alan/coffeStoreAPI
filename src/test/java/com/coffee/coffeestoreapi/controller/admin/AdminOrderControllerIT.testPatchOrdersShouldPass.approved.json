{
  "id": 10000003,
  "order_number": "ORD-1004",
  "orderer": "Tamas",
  "status": "PENDING",
  "discounts": [
    {
      "name": "25% off for orders over â‚¬12",
      "percentage": 25,
      "amount_in_cents": 450
    }
  ],
  "sub_total_price_in_cents": 1800,
  "total_price_in_cents": 1350,
  "currency": "EUR",
  "order_lines": [
    {
      "price_in_cents": 1300,
      "toppings": [
        {
          "name": "Chocolate Syrup",
          "price_in_cents": 500
        },
        {
          "name": "Hazelnut Syrup",
          "price_in_cents": 300
        }
      ]
    },
    {
      "price_in_cents": 500,
      "toppings": []
    }
  ],
  "created_at": "2023-10-04T11:20:00",
  "updated_at": "2023-10-04"
}